### PR TITLE
Add simple joi validation

### LINKED ISSUE #45: Config validation
We are adding more to the config setup. We currently have zero validation on formats and inputs which make the errors incredibly bad when making a mistake.

We should look at adding scheme validation.

Potentially we can use `joi` and document it.

### PR DESCRIPTION
**Pull request by pelzerim**
Adds simple joi validation to the cli.

closes #45

### CODE REVIEW

**Review by drubin — APPROVED** [review]

File: test/cli/schemas/replication-state-schema.test.js [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+  const validObject = { replicaSet: { replicaSetName, servers } }
+
+  const { error, value } = schema.validate(validObject)
+
+  t.is(error, undefined)
+  t.deepEqual(value, validObject)
+})
+
+test.serial('A invalid schema is detected as such', (t) => {
```
Comment (drubin) [inline test/cli/schemas/replication-state-schema.test.js:22]: You should add a test for name being missing. (this doesn't test all the cases)

File: lib/cli/schemas/replication-state-schema.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+const Joi = require('@hapi/joi')
+
+const schema = Joi.object({
+  replicaSet: Joi.object({
+    replicaSetName: Joi.string().required(),
+    servers:
+     Joi.array().min(1).required()
+       .items(Joi.object({
+         host: Joi.string().required(),
+         uri: Joi.string().uri()
```
Comment (drubin) [inline lib/cli/schemas/replication-state-schema.js:10]: https://hapi.dev/module/joi/api/?v=17.1.1#stringurioptions we can also validate mongo:// is the only schemee

File: lib/cli/schemas/replication-state-schema.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+       }))
+  }).required()
+}).unknown(true) // This allows other keys alongside "replicaSet" (argv adds private properties)
```
[the remark this answers is not in the document]
Comment (pelzerim) [reply lib/cli/schemas/replication-state-schema.js:13]: I don’t fully understand

[the remark this answers is not in the document]
Comment (drubin) [reply lib/cli/schemas/replication-state-schema.js:13]: We an inject all the settings to `argv.settings.repliaset` this way you can remove `unknown(true)` the reason I like this is that if someone makes a typo for an optional field the validation will pick it up. 


I also had this odd idea of having a config scheme being in a  single file/doc so the entire config is always validated with the same validation engine, not parts/sections. 

Think about how we add the `connection.uri` (from the user's PR) property to the in this method?

File: lib/cli/schemas/replication-state-schema.js (hunk 2 above)
Comment (drubin) [reply lib/cli/schemas/replication-state-schema.js:10]: btw joi can also default a value to a child value (this way we can remove  those hacks from our libs) 

But I  can't  find the link now

ie you can do `uri.defaults((obj) -> `mongodb://${object.host}`)` something like  this but  the syntax might be off

**Comment by drubin** [comment]
I would like to merge the users first and push this on top of that if possible as this MR is much easier to rebase.

**Pull request closed without merging**